File: src/algorithms/python/Word_Ladder.py
# -*- coding: utf-8 -*-

from Queue import Queue


class Solution:
    # @param beginWord, a string
    # @param endWord, a string
    # @param wordDict, a set<string>
    # @return an integer
    def ladderLength(self, beginWord, endWord, wordDict):
        if beginWord == endWord:
            return []
        alpha_list = [chr(ord('a') + i) for i in range(26)]
        wordDict.add(endWord)
        w_length = len(beginWord)
        vis = {beginWord: True}
        q = Queue()
        q.put((beginWord, 0))
        while not q.empty():
            top_word, top_cnt = q.get()
            for i in range(w_length):
                part1, part2 = top_word[:i], top_word[i + 1:]
                for ch in alpha_list:
                    if ch != top_word[i]:
                        tmp = part1 + ch + part2
                        if tmp in wordDict:
                            new_cnt = top_cnt + 1
                            if tmp == endWord:
                                return new_cnt + 1
                            else:
                                if tmp not in vis:
                                    vis[tmp] = new_cnt
                                    q.put((tmp, new_cnt))
        return 0


if __name__ == "__main__":
    solution = Solution()
    print solution.ladderLength(
        "hit", "cog", {"hot", "dot", "dog", "lot", "log"}
    )

File: src/algorithms/cpp/Word_Ladder.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <unordered_set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
private:
    string beginWord;
    map <string, set <string> > mp;
    vector <string> result_element;

public:
    int ladderLength(string beginWord, string endWord, unordered_set <string> &wordList) {
        this->beginWord = beginWord;
        int len = SZ(beginWord);
        map <string, int> vis;
        vector <char> alpha_list;
        queue <string> q;
        if (beginWord == endWord) {
            return 1;
        }
        vis[beginWord] = 1;
        wordList.insert(beginWord);
        wordList.insert(endWord);
        for (char ch = 'a'; ch <= 'z'; ++ ch)   alpha_list.push_back(ch);
        q.push(beginWord);
        while (!q.empty()) {
            string front = q.front();
            q.pop();
            for (int i = 0; i < len; ++ i) {
                string left = string(front.begin(), front.begin() + i);
                string right = string(front.begin() + i + 1, front.end());
                for (char &ch: alpha_list) {
                    if (ch == front[i])   continue;
                    string tmp_s = left + ch + right;
                    if (wordList.count(tmp_s) == 0)  continue;
                    int tmp_cnt = vis[front] + 1;
                    if (vis[tmp_s] != 0 && vis[tmp_s] < tmp_cnt)    continue;
                    vis[tmp_s] = tmp_cnt;
                    if (tmp_s == endWord) {
                        return vis[endWord];
                    }
                    else {
                        if (vis[endWord] == 0 || tmp_cnt < vis[endWord]) {
                            if (mp[tmp_s].count(front) == 0) {
                                mp[tmp_s].insert(front);
                                q.push(tmp_s);
                            }
                        }
                    }
                }
            }
        }
        return vis[endWord];
    }
};

int main() {
    Solution solution = Solution();
    string beginWord = "hit", endWord = "cog";
    unordered_set <string> wordList {"hot", "dot", "dog", "lot", "log"};
    cout << solution.ladderLength(beginWord, endWord, wordList) << endl;
    return 0;
}
